src/main.py

```diff
@@ -756,38 +757,45 @@ def _accept_switchover(self, lock_holder, previous_primary):
         if not self._can_do_switchover():
             return None
 
-        # WARNING: we shouldn't allow multiple hosts to enter this branch
-        if not self.zk.write(self.zk.SWITCHOVER_STATE_PATH, 'candidate_found', need_lock=False):
-            logging.error('Failed to state that we are the new primary candidate in ZK.')
+        switchover_candidate = self.zk.get(self.zk.SWITCHOVER_CANDIDATE)
+        if switchover_candidate is None:
+            logging.info('Waiting for primary to choose switchover candidate...')
             return None
 
-        #
-        # All checks are done. Waiting for primary shutdown, acquiring the lock in ZK,
-        # promoting and writing last switchover timestamp to ZK.
-        #
+        logging.info('Switchover candidate is: %s', switchover_candidate)
+        if switchover_candidate != helpers.get_hostname():
+            logging.info('Current host is not the candidate, turning to the new master')
+            return self._return_to_cluster(switchover_candidate, 'replica')
+
+        if self.zk.get(self.zk.SWITCHOVER_STATE_PATH) == 'candidate_failover':
+            logging.warning('Switchover state is aready candidate_failover, proceeding to failover')
+        else:
+            logging.info('Current host is the candidate and ready, signaling primary...')
+            if not self.zk.write(self.zk.SWITCHOVER_STATE_PATH, 'candidate_ready', need_lock=False):
+                logging.error('Failed to state that we are the new primary candidate in ZK.')
+                return None
+
         limit = self.config.getfloat('global', 'postgres_timeout')
-        # Current primary is lock holder. Otherwise consider last primary as current.
-        current_primary = lock_holder or previous_primary
-        if current_primary is not None and not helpers.await_for(
-            lambda: self._check_primary_is_really_dead(current_primary), limit, 'primary is down'
-        ):
-            return None
 
-        # Wait switchover_master_shut state only if current primary is alive, i.e. lock holder exists.
+        # Wait candidate_failover state only if current primary is alive, i.e. lock holder exists.
         if lock_holder is not None and not helpers.await_for(
```

Comment:
if old master became dead before setting candidate_failover - we will wait here, but it makes no sense. Mb it would be better if we will check if old master release leader_lock? Looks like we don't care if old master alive (and they set candidate_failover and release leader lock) or dead (just release leader lock) here.

Reply:
yes..  step with `candidate_failover` really looks redundant